### PR TITLE
haskellPackages.*: unbreak glirc/matterhorn/QuickCheck/etc

### PR DESCRIPTION
**Pull request by Kiwi**
###### Motivation for this change
they were broken or marked broken and at least two people use a few of them :)

###### Things done
i fixed them and marked them not broken
tested glirc and matterhorn (now on macos as well)

mattermost-api{-qc} weren't even actually broken? *shrug*

am trying to compile them for macos but that might take a while... (glirc worked waiting on matterhorn) and doing nixpkgs-review on aarch64 because why not? (found out why; it doesn't work!)
- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 1]
```diff
@@ -324,6 +325,7 @@ self: super: {
   http-link-header = dontCheck super.http-link-header; # non deterministic failure https://hydra.nixos.org/build/75041105
   ihaskell = dontCheck super.ihaskell;
   influxdb = dontCheck super.influxdb;
+  irc-core = doJailbreak super.irc-core;
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix:328]: Could you report this upstream and link to the issue here in the code?

This is so that we have some way to track when the `doJailbreak` can be removed.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 2]
```diff
@@ -335,7 +337,7 @@ self: super: {
     then dontCheck super.math-functions # "erf table" test fails on Darwin https://github.com/bos/math-functions/issues/63
     else super.math-functions;
   matplotlib = dontCheck super.matplotlib;
-
+  matterhorn = doJailbreak super.matterhorn;
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix:340]: Same thing here.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 3]
```diff
@@ -826,6 +832,17 @@ self: super: {
   # https://github.com/aisamanra/config-ini/issues/12
   config-ini = dontCheck super.config-ini;
 
+  config-value = (doJailbreak (overrideCabal super.config-value (drv: {
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix:835]: Same thing here.

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 4]
```diff
@@ -826,6 +832,17 @@ self: super: {
...
+    buildTools = [
+      (super.callHackageDirect
+        {
+          pkg = "happy";
+          ver = "1.19.12";
+          sha256 = "0n1ri85hf1h9q5pwvfnddc79ahr9bk7hz8kirzrlyb81qzc3lpc9";
+        }
+        { })
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix:843]: What is the problem with `happy`?  Does `config-value` need an older or a newer version?

In general, things using IFD won't work in nixpkgs because of hydra, so you won't be able to use `callHackageDirect` for this.

The `extra-packages` section in `configuration-hackage2nix.yaml` lets you define new packages that our Haskell stuff will generate and provide if you need some other version of `happy`.

Comment (Kiwi) [reply pkgs/development/haskell-modules/configuration-common.nix:843]: yeah I thought maybe I wasn't allowed to use that in nixpkgs. :( I suspected the extra-packages is what I should use. too. (I asked on IRC but didn't get a reply *shrug*)

```
  build-tool-depends:
    alex:alex   ^>= 3.2.4,
    happy:happy ^>= 1.19,
```

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 1 above)
Comment (Kiwi) [reply pkgs/development/haskell-modules/configuration-common.nix:328]: I can and am about to start making them. I have a problem in that I'm really not sure what I'm supposed to say; I could really use some guidance or a template or even another issue someone's made that's a good example to follow (there's probably another time I've asked that you explained it I could link to I will try to find) I appreciate all the work you do. [I will paint you any stickers you want.](https://mostlyabsurd.com/files/stickers-for-all.html) I'm on the IRC and Discord. let me know what you'd like I'm still making them... I think I tried getting your attention on Twitter a while ago by adding you to a group but I don't know if it worked :laughing: . and all of [my contact information are here](https://www.mostlyabsurd.com/contact/#socialize-with-me) Anyway, as far as the issues I'm supposed to make go... what I want to say is far too blunt and that's just how I am. (I really don't *want* to nor *will* I say this but it's how I think a lot of people unfamiliar with NixOS experience it. I do know a few people that are bewildered by it though even after I try and explain, which is probably me not being good at explaining :stuck_out_tongue:  . anyway that's why I could use help. >.>

>  "we use stackage on NixOS (@cdepillabout and I know part of the reason we do is there are too many versions on hackage for NixOS to have them all. doesn't make it any less annoying) and you publish to hackage and there's probably a reason you don't use stackage I know a lot of people don't like it (I was already not a fan before I even heard of NixOS and I've only come to dislike it more) but I am here to suggest you do anyway for all 2 (idk how many people use it I hope there are more) people that use glirc on NixOS who don't want to use overlays so I'm cluttering your issue tracker; also please update your version bounds ASAP because while you don't use NixOS and only support cabal install where what you have works fine the fact that this doesn't work gives a bad impression about Nix"

 @glguy is a friend and he's pretty quick about fixing the bounds if I ask (sometimes he's even started before I ask). but most of these people I don't know at all :frowning_face: and what happens if they don't want to or can't? can they just be like permanent? and If they need updated to keep working it's not like I'm already updating them to keep them working >.> they're there most of the time anyway.

Comment (Kiwi) [reply pkgs/development/haskell-modules/configuration-common.nix:328]: [found your comment](https://github.com/NixOS/nixpkgs/pull/92618#discussion_r451303303)

btw here's one of the haskell stickers that I painted

and NixOS

I'm giving like at least 144 away (but so far there are a lot more stickers than there are people to give them...) :cry:  (they're really cool and take a long time and, well, they aren't cheap lol. but people really like them) Everyone here does so much good (unpaid, often underappreciated) work that it's the least I can do (it's all explained in the blog...)

@peti you can have some too even though you seem to make most of the commits that make me fix them :stuck_out_tongue_winking_eye: but I don't think it'd work much at all without you so you have that in your favor :wink: 

really anyone that wants one just lmk

File: pkgs/development/haskell-modules/configuration-common.nix (hunk 4 above)
Comment (glguy) [reply pkgs/development/haskell-modules/configuration-common.nix:843]: I'll look into fixing the bound on happy to accept 1.20

Comment (Kiwi) [reply pkgs/development/haskell-modules/configuration-common.nix:843]: I clicked over to your repo to file an issue after my last comment here and I had a notification :rofl: :smile: @cdepillabout like I said glguy is on point! :heart:

File: pkgs/development/haskell-modules/configuration-common.nix [hunk 5]
```diff
@@ -335,7 +338,7 @@ self: super: {
     then dontCheck super.math-functions # "erf table" test fails on Darwin https://github.com/bos/math-functions/issues/63
     else super.math-functions;
   matplotlib = dontCheck super.matplotlib;
-
+  matterhorn = doJailbreak super.matterhorn; # this is needed until the end of time
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix:341]: > this is needed until the end of time

Is there some reason why this can't be reported upstream and get the author to fix the bounds in the .cabal file?

For instance, when trying to build `matterhorn` without `doJailbreak`, Cabal complains about:

```console
$ nix-build -A haskellPackages.matterhorn
...
configuring matterhorn-50200.11.0...
CallStack (from HasCallStack):
  $, called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:1024:20 in Cabal-3.2.0.0:Distribution.Simple.Configure
  configureFinalizedPackage, called at libraries/Cabal/Cabal/Distribution/Simple/Configure.hs:477:12 in Cabal-3.2.0.0:Distribution.Simple.Configure
  configure, called at libraries/Cabal/Cabal/Distribution/Simple.hs:625:20 in Cabal-3.2.0.0:Distribution.Simple
  confHook, called at libraries/Cabal/Cabal/Distribution/Simple/UserHooks.hs:65:5 in Cabal-3.2.0.0:Distribution.Simple.UserHooks
  configureAction, called at libraries/Cabal/Cabal/Distribution/Simple.hs:180:19 in Cabal-3.2.0.0:Distribution.Simple
  defaultMainHelper, called at libraries/Cabal/Cabal/Distribution/Simple.hs:116:27 in Cabal-3.2.0.0:Distribution.Simple
  defaultMain, called at Setup.hs:2:8 in main:Main
Setup: Encountered missing or private dependencies:
aeson >=1.2.3.0 && <1.5,
aspell-pipe ==0.4.*,
skylighting-core ==0.8.*,
strict ==0.3.*,
vty ==5.31.*
```

You can check which versions of these things we have in nixpkgs:

```console
$ nix-broken -A haskellPackages.aeson
/nix/store/kdlv4s53whzllsvv5xqnn15f6lv9z7zi-aeson-1.5.4.1
```

So the `matterhorn.cabal` file would need the `aeson` version bound bumped to probably `aeson  < 1.6`.  The other dependencies should have similar bumps made.

It looks like there are other PRs making these bumps, so the author is probably okay with these types of things: https://github.com/matterhorn-chat/matterhorn/pull/487

Comment (Kiwi) [reply pkgs/development/haskell-modules/configuration-common.nix:341]: lol yes, because I was being funny and didn't have an issue to put there yet. I have too many things going on and get distracted by shiny things. I was in the mood to update the PR with the changes for glirc before I was in the mood to file an issue; I hate making issues. "No Issues! only PRs!" I have made one now

Comment (cdepillabout) [reply pkgs/development/haskell-modules/configuration-common.nix:341]: @Kiwi could you update the code here to link to the issue so that the nixpkg Haskell maintainers can track it?

Comment (Kiwi) [reply pkgs/development/haskell-modules/configuration-common.nix:341]: yeah I am about to do that now

**Review by Kiwi — COMMENTED** [review]
edit: this was supposed to show up as being related to this; https://github.com/NixOS/nixpkgs/pull/104770/commits/34ae98d96dd0524ae306f287bfecb348d965687b and it's bs that I can't ask myself to make changes @github

Should these have been in two different commits so that ofborg could build both of them? (it does build them when I name the commit properly, which I think it is, doesn't it? or not because haskell. or maybe not because of multiple commits? it only seems to check one...)

I guess If I had put them in the right order, assuming they do get built, it would still be fine and have to build them both because config-schema depends on config-value so would build it. That was my plan but I had it the other way around and thought config-value depended on config-schema.

Anyway I have built everything on darwin, nixos, and aarch64 nixos. so it should be fine.

File: pkgs/development/haskell-modules/configuration-hackage2nix.yaml [hunk 6]
```diff
@@ -2694,6 +2694,15 @@ package-maintainers:
...
   kiwi:
+    - config-schema
+    - config-value
+    - glirc
+    - irc-core
+    - matterhorn
+    - mattermost-api
+    - mattermost-api-qc
+    - QuickCheck_2_14_2
+    - random_1_2_0
```
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-hackage2nix.yaml:2705]: This PR looks good, although since these packages will probably get updated at some point, I'm not sure if it would be okay to keep them here.

I'll remove them just to be safe.

```suggestion
```

**Comment by cdepillabout** [comment]
Thanks for these changes.  This looks good so I am merging in.

---------------------------------------

I did have one question though just out of curiosity.  In https://github.com/matterhorn-chat/matterhorn/issues/679 you said that you don't actually use matterhorn.  However if I remember correctly, you're always the one that updates/fixes it in nixpkgs.  Why do you fix it if you don't use it?

**Comment by Kiwi** [comment]
gotta pump up my PR numbers if I ever want to get commit bit! :smiling_imp: 

earlier on Discord:
> @cdepillabout can we just assume that anything i fix is going to require the overrides forever ;_;
> not that i don't like inflating my PR count with essentially the same PRs every few weeks...
> "so kiwi what makes you think you're qualified to be a nixpkgs committer?" "well I've made 800 prs to these 3 packages on nixpkgs" "..." "can haz commit?"

More seriously though @evils uses it and I'm sure other people do too and he does most of the work on Kicad so it's the least I could do for him. and who knows, maybe I'll want to use it someday and I won't even have to fix/package it first when I do! :smile: . I like to make sure things work and are available in nixpkgs and that goes a lot smoother if there's at least, and preferably more than, one maintainer listed. Not all maintainers are even that active. or around at all. So when I find something that's been broken for a while or I am interested in using or if I just want to take on some of the work (hi @florianjacob) to get things done quicker (I have a lot of free time and nothing to do) I'll fix/upgrade/package it and put my name down. If nobody else has done it by the time I do it then who knows if or when something will work again. (like the bitwarden-desktop client, a few znc plugins i needed to replace my freebsd server, MLV.APP, and simplenote I wanted to try and had to package) So if I can figure it out I do it.. "No Issues, only PRs!" :stuck_out_tongue: . ...I still need to make a few of those work on darwin now that I have a builder...

...Sometimes I'll fix something that breaks in nixos-unstable and come to make the PR and there is already one, other times it's been weeks or months without attention. I think at least half a dozen people tried getting ocrmypdf packaged before I did... it took me a few months and I've never used it other than to test but I know other people do because they file issues when it breaks. lol (I have used the program but the one PDF I needed it for I processed with the brew package on macos while I was getting used to NixOS in a VM on my laptop and an old desktop someone was throwing away). That's also how I ended up having [nixops-digitalocean](https://github.com/nix-community/nixops-digitaloceanl) :cry: It got axed when nixops 2 went with the plugin system and nobody had converted it to a plugin for a few months and so I gave it a shot... anyway...

tl;dr: somebody's gotta do it and if I can figure it out and have time (I usually do) then it might as well be me because if I don't maybe no-one will and it gives me something to do and a way to help. same reason I'm making all of these stickers to give maintainers (hi @jonringer)

**Comment by cdepillabout** [comment]
Well thanks for spending the time to get this stuff working!
